tango/integrations/gs/workspace.py

```diff
@@ -0,0 +1,225 @@
+        with concurrent.futures.ThreadPoolExecutor(
+            max_workers=self.NUM_CONCURRENT_WORKERS,
+            thread_name_prefix="GSWorkspace._get_run_from_dataset()-",
+        ) as executor:
+            step_info_futures = []
+            for unique_id in steps_info.values():
+                step_info_futures.append(executor.submit(self.step_info, unique_id))
+            for future in concurrent.futures.as_completed(step_info_futures):
+                step_info = future.result()
+                assert step_info.step_name is not None
+                steps[step_info.step_name] = step_info
+
+        return Run(name=run_entity.key.name, start_date=run_entity["start_date"], steps=steps)
+
+    def registered_runs(self) -> Dict[str, Run]:
+        import concurrent.futures
+
+        runs: Dict[str, Run] = {}
+
+        with concurrent.futures.ThreadPoolExecutor(
+            max_workers=self.NUM_CONCURRENT_WORKERS,
+            thread_name_prefix="GSWorkspace.registered_runs()-",
+        ) as executor:
+            run_futures = []
+            for run_entity in self._ds.query(kind="run").fetch():
+                run_futures.append(executor.submit(self._get_run_from_entity, run_entity))
+            for future in concurrent.futures.as_completed(run_futures):
+                run = future.result()
+                if run is not None:
+                    runs[run.name] = run
+
+        return runs
+
+    def registered_run(self, name: str) -> Run:
+        err_msg = f"Run '{name}' not found in workspace"
+
+        run_entity = self._ds.get(key=self._ds.key("run", name))
+        if not run_entity:
+            raise KeyError(err_msg)
+
+        run = self._get_run_from_entity(run_entity)
+        if run is None:
+            raise KeyError(err_msg)
+        else:
+            return run
+
+    def step_info(self, step_or_unique_id: Union[Step, str]) -> StepInfo:
+        unique_id = (
+            step_or_unique_id if isinstance(step_or_unique_id, str) else step_or_unique_id.unique_id
+        )
+        step_info_entity = self._ds.get(key=self._ds.key("stepinfo", unique_id))
+        if step_info_entity:
+            step_info_bytes = step_info_entity["step_info_dict"]
+            step_info = StepInfo.from_json_dict(json.loads(step_info_bytes))
+            return step_info
+        else:
+            if not isinstance(step_or_unique_id, Step):
+                raise KeyError(step_or_unique_id)
+            step_info = StepInfo.new_from_step(step_or_unique_id)
+            self._update_step_info(step_info)
+            return step_info
+
+    def _update_step_info(self, step_info: StepInfo):
+
+        step_info_entity = self._ds.entity(
+            key=self._ds.key("stepinfo", step_info.unique_id),
+            exclude_from_indexes=("step_info_dict",),
+        )
+
+        # We can store each key separately, but we only index things that are useful for querying.
+        # TODO: do we want to index any other step_info keys?
```

Comment:
Can't think of anything, but maybe it will become apparent when we merge the UI stuff?